### PR TITLE
atomic metadata call for PRC

### PR DESCRIPTION
**Pull request by d-w-moore**
To address #244 .
This is a WIP and should be adjusted to accommodate a more succinct syntax.
We start by just jamming in the JSON directly ...
```python
  ses=get_session()
  reqs = {
    "entity_name": "/tempZone/home/rods/thing.cc",
    "entity_type": "data_object",
    "operations": [
      {
        "operation": "remove",
        "attribute": "a1",
        "value": "v1"
      },
      {
        "operation": "add",
        "attribute": "a2",
        "value": "v2",
        "units": "u2"
      }
    ]
  }
  ses.metadata.atomic( reqs )
```
It possibly should look more like ...
```
sys.metadata.atomic ( [ 'remove', iRODSMeta( 'a1','v1') ] ,   [ 'add', iRODSMeta( 'u2','v2',u2') ] , #...
)
```
or even:
```
[data|resc|coll|user].metadata.atomic ( [...],[...],...)
```

### CODE REVIEW

**Comment by d-w-moore** [comment]
Now just need a test or two....

**Comment by korydraughn** [comment]
> It possibly should look more like ...
> 
> ```
> sys.metadata.atomic ( [ 'remove', iRODSMeta( 'a1','v1') ] ,   [ 'add', iRODSMeta( 'u2','v2',u2') ] , #...
> )
> ```
> 
> or even:
> 
> ```
> [data|resc|coll|user].metadata.atomic ( [...],[...],...)
> ```

Just curious, why use an array to specify the add/remove and the AVU instead of something else? An array feels too loose IMO. Just so there is no confusion, I'm asking about the following proposed syntax:
```
sys.metadata.atomic ( [ 'remove', iRODSMeta( 'a1','v1') ] ,   [ 'add', iRODSMeta( 'u2','v2',u2') ] , #... )
```

**Comment by d-w-moore** [comment]
Would an object hierarchy be better I wonder ...
Eg obj.metadata.atomic(  iRodsAtomicAVURemove ( 'a','v','u' ), iRodsAtomicAVUAdd( 'a2','v2','u2' ), # ... etc 
)
where `iRODSAtomicAVU{Remove,Add}` have a common base class `iRODSAtomicAVUOperation`?

**Comment by trel** [comment]
i like where this is going... but want to avoid making everything require 1000 characters...

we also need to consider/appreciate the ordered nature of the atomic add/remove operations...   so perhaps putting all the operations in an array is good too?

**Comment by d-w-moore** [comment]
We could have this:
```

 obj.metadata.atomic (
         { 'operation':'remove', 
           'values':   [ iRODSMeta(...), iRODSMeta(...), ...]
         }, 
         { 'operation':'add', 
           'values':   [ iRODSMeta(...), iRODSMeta(...), ...]
         }, 
              #...
)
```
That might be closer to the spirit of the original thing.

```

**Comment by trel** [comment]
would that insist that all the removes happen before all the adds?

what if we want them interleaved?

i'm not convinced having multiple `iRODSMeta`'s in an array outweighs the effort to expand the mental model.

original thing has simple array of these...  'operation', 'a', 'v', 'u'

operations are ordered, singular, and easy to read later

**Comment by korydraughn** [comment]
I like these ideas. As for your last idea, remember the API plugin does not accept many AVUs for one **operation** (e.g. `{"operation": "remove", "values": [...]}`. The **operation** would have to be mapped to each AVU in the **values** array.

I definitely like the idea of using specific types to constrain the interface tho.

**Comment by korydraughn** [comment]
What if the atomic function just took JSON or a dictionary directly? Python has a JSON library already. We could then provide utility functions (if needed) that make it easy to construct the JSON instead of having the function enforce a particular interface.

**Comment by trel** [comment]
i'm in for JSON

**Comment by d-w-moore** [comment]
I am not sure PRC programmers would prefer the JSON over iRODSMeta objects.
FIrstly, it won't  ever be exactly the same JSON used by the raw irods interface. (The object path/name and type are already abstracted out, as you can see.) So we'll have a 'slightly different' JSON at best.
Secondly iRODSMeta objects are part-and-parcel to how PRC deals with metadata. If (as a programmer using PRC) you can build a data-structure directly using iRODSMeta objects (and you can), maybe those should figure into the atomic metadata function interface too?  Easier for the PRC user than translating them ad-hoc to JSON structures, or requiring the user to write his own routines to do the job.

**Comment by trel** [comment]
so... we're back to...

```
obj.metadata.atomic( [ 'remove', iRODSMeta( 'a1','v1') ],
                     [ 'add',    iRODSMeta( 'u2','v2',u2') ],
                     # ...
                   )
```

and the operation order is positional and honored?

and `atomic()` will complain if it receives any operation other than `[add|remove]`.

**Comment by d-w-moore** [comment]
A simplifying compromise might be this...
```
  obj.metadata.atomic (  { } , { } , ... )
```
Where each dictionary `{ }` passed in is simply:
```json
{
   "operation" :   "add_or_remove"
   "value" : a_tuple_or_iRODSMeta_object
}
```

**Comment by korydraughn** [comment]
I'm for the dictionary approach. It is flexible and clear. The dict could be replaced by a class that simply holds a string and an iRODSMeta object (or tuple) if you wanted to constrain the interface more. That way, people could do things like:
```python
op = AtomicOp()
op.operation = 'remove'
op.avu.attribute = ...
op.avu.value = ...
op.avu.units = ...
# Where "avu" is an iRODSMeta object.

sys.metadata.atomic(op, ...)
```
I'm totally fine with either approach.

I'm against the array-based approach (i.e. `['add', IRODSMeta(...) ]`). This may lead to devs thinking they can pass more objects to the array.

**Comment by d-w-moore** [comment]
That's fair, will see what I can do with that.

**Comment by trel** [comment]
looking good.   can we get an example with both `a,v` and `a,v,u` added to README.md?

File: irods/test/meta_test.py [hunk 1]
```diff
@@ -29,13 +29,76 @@ def setUp(self):
...
+        m = iRODSMeta( "attr_244","value","units")
+        self.obj.metadata.add(m)
+        meta = self.sess.metadata.get(DataObject, self.obj_path)
+        self.assertEqual(len(meta), 1)
+        self.assertTrue(AVUs_Equal(m,meta[0],AVU_As_Tuple))
+        self.obj.metadata.apply_atomic_operations(                                  # remove original AVU and replace
+           AVUOperation(operation="remove",avu=m),                                  #   with two altered versions
+           AVUOperation(operation="add",avu=iRODSMeta(m.name,m.value,"units_244")), # (one of them without units) ...
+           AVUOperation(operation="add",avu=iRODSMeta(m.name,m.value))
+        )
```
Comment (korydraughn) [inline irods/test/meta_test.py:99]: If someone generated a list of `AVUOperation`s dynamically and wanted to pass that list to `metadata.apply_atomic_operations`, could they do that?

Can this function take an array/list as well?

Comment (korydraughn) [reply irods/test/meta_test.py:99]: I just noticed lines 63 and 64 above. Is that how someone would achieve what I just asked about?

**Comment by d-w-moore** [comment]
Working on docs for README right now...

File: irods/test/meta_test.py (hunk 1 above)
Comment (d-w-moore) [reply irods/test/meta_test.py:99]: Yes, those lines show the general strategy.  Distilled down, its like this:
```
def f( x, y ) : print (x, y)
def g( *z ) : print (z)
>>> f(3,4) # =>  prints (3,4)
>>> g(3,4) # => prints (3,4)
>>> b = [2,4,6,8] # or any computed list or tuple
>>> g(*b) # => prints out b, as a tuple
```

Comment (d-w-moore) [reply irods/test/meta_test.py:99]: And if you look at the function definitions for `apply_atomic_operations`, you can see they're variadic by virtue of taking a final `*parameter` where that `parameter` collects all "final" arguments being applied in the function call , however many,  into a tuple.

File: README.rst [hunk 2]
```diff
@@ -277,6 +277,34 @@ of the (type of) catalog object they once annotated:
...
+So, for example, if 'obj' is a handle to an object in the iRODS catalog, we can send an arbitrary number of AVUOperation
+instances to be executed together as one indivisible operation:
+
+>>> from irods.meta import iRODSMeta, AVUOperation
+>>> obj.metadata.apply_atomic_operations( AVUOperation(operation='remove', avu=iRODSMeta('a1','v1','these_units')),
+...                                       AVUOperation(operation='add', avu=iRODSMeta('a2','v2','those_units')),
+...                                       AVUOperation(operation='remove', avu=iRODSMeta('a3','v3')) # , ...
+... )
+
+The list of AVUOperation objects will effected in the order given, so that a "remove" followed by an "add" of the same AVU
```
Comment (trel) [inline README.rst:296]: 'be effected' -> 'be applied'

File: README.rst [hunk 3]
```diff
@@ -277,6 +277,34 @@ of the (type of) catalog object they once annotated:
...
 
 
+Atomic operations on metadata
+-----------------------------
+
+With release 4.2.8 of iRODS, the atomic metadata API was introduced to allow a group of metadata add and remove
+operations to be performed transactionally, within a single call to the server.  This capability can be leveraged in
+version 0.8.6 of the PRC.
+
+So, for example, if 'obj' is a handle to an object in the iRODS catalog, we can send an arbitrary number of AVUOperation
```
Comment (trel) [inline README.rst:287]: please additionally note that this works for objects, collections, users, and resources.

**Review by trel — COMMENTED** [review]
i think this may be it... anything else left to do?

File: irods/test/meta_test.py (hunk 1 above)
Comment (trel) [reply irods/test/meta_test.py:99]: i think this is covered.   good stuff.

**Comment by trel** [comment]
consider commit message of 'add capability to add/remove atomic metadata'

**Comment by d-w-moore** [comment]
Think that's it... test suite passes on Python2 and 3

**Comment by trel** [comment]
excellent.   merging.

**Pull request merged**